### PR TITLE
wallet: Make a tr() descriptor by default

### PR DESCRIPTION
**Pull request by achow101**
Make a `tr()` descriptor by default in descriptor wallets so that users will be able to make and use segwit v1 bech32m addresses.

### CODE REVIEW

**Comment by Sjors** [comment]
The first couple of commits can be merged earlier with a separate PR.

Also, it'd be nice to have this on Signet, so explictly check for that chain and then make a followup to do it on testnet and mainnet?

How do you want to go about upgrading existing descriptor wallets? Given that it's still marked experimental, having some manual RPC / wallet tool command would be fine I think. But with future soft-forks such an upgrade should be automatically, just like how we added Segwit addresses to existing wallets.

When I tried getting the master HD key to derive some new private keys in #22341 I got thoroughly confused. It might make more sense to store the seed in the main wallet payload rather than in each SPKman.

**Comment by vansh143** [comment]
What is this bro.

**Review by lesliebranson — APPROVED** [review]

**Comment by achow101** [comment]
The first commit was split to #22512

**Comment by michaelfolkson** [comment]
Concept ACK

I misunderstood what the "default" aspect of this PR was referring to. As explained to me in the July 30th Core [wallet meeting](https://gnusha.org/bitcoin-core-dev/2021-07-30.log) there is no universal "default" descriptor in the wallet, just a default descriptor per address type. This PR is allowing a Taproot descriptor be constructed (and will be merged post Taproot activation) rather than constructing a Taproot descriptor as a universal "default" descriptor.

**Comment by Sjors** [comment]
I made a [branch](https://github.com/Sjors/bitcoin/tree/2021/10/tr-desc-and-bech32m-gui) that combines this PR with #22260, so you get bech32m addresses in the GUI for new descriptor wallets. It seems to work fine.

Given that this PR has to wait for Taproot, I suggest splitting off all commits except 449ce29.

Suggest rewording the PR title so it's more clear the tr() is created, but does not become the default. We could additionally mention in the release notes that creating a taproot address is possible using the console: `getnewaddress label bech32m`, and explain that the GUI doesn't do this automatically yet due to lack of broad bech32m support.

**Review by maflcko — COMMENTED** [review]
Concept ACK. Only took a look at the tests and had some questions.

**Comment by MarcoFalke** [comment]
nit, if you rebase and push:

```diff
diff --git a/src/wallet/test/fuzz/notifications.cpp b/src/wallet/test/fuzz/notifications.cpp
index 252832785b..0811acca13 100644
--- a/src/wallet/test/fuzz/notifications.cpp
+++ b/src/wallet/test/fuzz/notifications.cpp
@@ -68,9 +68,6 @@ struct FuzzedWallet {
     CScript GetScriptPubKey(FuzzedDataProvider& fuzzed_data_provider)
     {
         auto type{fuzzed_data_provider.PickValueInArray(OUTPUT_TYPES)};
-        if (type == OutputType::BECH32M) {
-            type = OutputType::BECH32; // TODO: Setup taproot descriptor and remove this line
-        }
         CTxDestination dest;
         bilingual_str error;
         if (fuzzed_data_provider.ConsumeBool()) {

**Comment by MarcoFalke** [comment]
Adjusted title again for 🥕

**Comment by Sjors** [comment]
tACK f04337ebae26048489e4992992d13a59cfa03ec7

**Comment by sipa** [comment]
Concept ACK

**Comment by laanwj** [comment]
> How do you want to go about upgrading existing descriptor wallets? Given that it's still marked experimental, having some manual RPC / wallet tool command would be fine I think. But with future soft-forks such an upgrade should be automatically, just like how we added Segwit addresses to existing wallets.

Same question (couldn't find an answer in this PR)—does this add the descriptor to existing wallets, or only to new ones? If not, what is the upgrade command, and should it be added to the release notes?

**Comment by achow101** [comment]
> Same question (couldn't find an answer in this PR)—does this add the descriptor to existing wallets, or only to new ones? If not, what is the upgrade command, and should it be added to the release notes?

This PR only applies to newly created descriptor wallets. Current wallet will need to be upgraded through some method that I haven't quite implemented yet. It will probably be some extension of `upgradewallet` and make use of #23417.

**Comment by MarcoFalke** [comment]
why not https://github.com/bitcoin/bitcoin/pull/22364#issuecomment-961802212 ?

File: src/script/signingprovider.cpp [hunk 1]
```diff
@@ -208,5 +208,15 @@ CKeyID GetKeyForDestination(const SigningProvider& store, const CTxDestination&
             }
```
Comment (MarcoFalke) [inline src/script/signingprovider.cpp:208]: Does the comment in this function need an update?

**Comment by achow101** [comment]
> why not [#22364 (comment)](https://github.com/bitcoin/bitcoin/pull/22364#issuecomment-961802212) ?

Forgot, done now.

File: src/script/signingprovider.cpp (hunk 1 above)
Comment (achow101) [reply src/script/signingprovider.cpp:208]: Done

**Comment by MarcoFalke** [comment]
Concept ACK 4868c9f1b39f03adee0009cd41d96598b43e8b78

**Comment by Sjors** [comment]
re-utACK 4868c9f1b39f03adee0009cd41d96598b43e8b78

**Review by meshcollider — COMMENTED** [review]
Concept + code review ACK 4868c9f1b39f03adee0009cd41d96598b43e8b78

**Comment by gruve-p** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/22364/commits/4868c9f1b39f03adee0009cd41d96598b43e8b78

**Comment by fanquake** [comment]
This was part of 23.0 and got a release note. Removing "Needs release note".
